frontend/test/metabase/scenarios/collections/collection-types.cy.spec.js

```diff
@@ -68,16 +87,56 @@ describeWithoutToken("collection types", () => {
   });
 
   it("should not display official collection icon", () => {
-    cy.createCollection({
-      name: "Official Collection Test",
-      authority_level: "official",
-    });
-    cy.visit("/collection/root");
-    cy.findByText("Official Collection Test").click();
-    cy.findByTestId("official-collection-marker").should("not.exist");
+    testOfficialBadgePresence(false);
   });
 });
 
+function testOfficialBadgePresence(expectBadge = true) {
+  cy.createCollection({
+    name: COLLECTION_NAME,
+    authority_level: "official",
+  }).then(response => {
+    const { id: collectionId } = response.body;
+    cy.createQuestion({
+      name: "Official Question",
+      collection_id: collectionId,
+      query: TEST_QUESTION_QUERY,
+    });
+    cy.createDashboard("Official Dashboard", { collection_id: collectionId });
+    cy.visit(`/collection/${collectionId}`);
+  });
+
+  // Collections page
+  cy.findByTestId("official-collection-marker").should(
+    expectBadge ? "exist" : "not.exist",
+  );
+  assertSidebarIcon(COLLECTION_NAME, expectBadge ? "badge" : "folder");
```

Comment:
Would encourage also extracting each of this blocks into a carefully-named function.


So one for Collections page, another for Dashboard, Question, Search

Reply:
Sounds great to me, will do! Thank you

Reply:
@gusaiani tried to extract those, but now I'm thinking if it's worth it. Please take a look at the points I'm uncertain about

**Dashboard and question pages**

For both pages, the actual assertion is already extracted into a single `assertHasCollectionBadge` function. Other lines are just about moving between pages. Not sure they're worth being extracted into a different function. If you rearrange them later, the test will break as it relies on the order of pages

```js
// Dashboard Page
cy.findByText("Official Dashboard").click(); // opens dashboard page from the collection page
assertHasCollectionBadge(expectBadge); // asserts dashboard page has badge

// Cypress in on the dashboard page now

// Question Page
cy.findByText(COLLECTION_NAME).click(); // Goes to the collection from dashboard's CollectionBadge
cy.findByText("Official Question").click(); // Opens a question from the collection page
assertHasCollectionBadge(expectBadge); // asserts question page has badge
```

**Search**

The search test relies on every items' name starting with "Official". In that way, Cypress can do only a single search query that would output a collection, a question, and a dashboard. If we extract it in a separate function, we would either need to hardcode them ( 😟 ) or accept them as arguments (but the test would break if you change item names)

```js
cy.get(".Nav")
    .findByPlaceholderText("Search…")
    .as("searchBar")
    .type("Official");

  cy.findByTestId("search-results-list").within(() => {
    assertSearchResultBadge(COLLECTION_NAME, {
      expectBadge,
      selector: "h3",
    });
    assertSearchResultBadge("Official Question", { expectBadge });
    assertSearchResultBadge("Official Dashboard", { expectBadge });
  });
``

Reply:
@kulyk To understand this myself, I made this PR.

https://github.com/metabase/metabase/pull/17151

It's pointed towards this branch.

The tests run all right it seems.

If you see any value in them, use them. If not, I will survive :) <3

Reply:
Thank you so much for opening the PR @gusaiani! I really like the test became more readable, though I'm still a bit concerned about using hardcoded item names and about fragile function calls order

@nemanjaglumac could you please take a look at our conversion here and recommend what's the preferred approach?

Reply:
I don't mind extracting the logic into helper functions, but if you do that please do not hard code collection names, ids and such. Ideally, pass them as named arguments (if there are multiple args needed).

```javascript
function changeCollectionFromOfficialToRegular(collectionName=COLLECTION_NAME) {
  cy.icon("pencil").click();
  cy.findByText("Edit this collection").click();
  modal().within(() => {
    setOfficial(false);
    cy.button("Update").click();
  });
  cy.findByTestId("official-collection-marker").should("not.exist");
  assertSidebarIcon(collectionName, "folder");
}
```
or even better - do not couple these two things

```javascript
function changeCollectionFromOfficialToRegular() {
  cy.icon("pencil").click();
  cy.findByText("Edit this collection").click();
  modal().within(() => {
    setOfficial(false);
    cy.button("Update").click();
  });
  cy.findByTestId("official-collection-marker").should("not.exist");
}

changeCollectionFromOfficialToRegular();
assertSidebarIcon(COLLECTION_NAME, "folder");
```

I'd probably extract the logic for `editCollection` because it'll be used all over the place
```javascript
function editCollection() {
  cy.icon("pencil").click();
  cy.findByText("Edit this collection").click();
}

function changeCollectionFromOfficialToRegular() {
  modal().within(() => {
    setOfficial(false);
    cy.button("Update").click();
  });
  cy.findByTestId("official-collection-marker").should("not.exist");
}

editCollection();
changeCollectionFromOfficialToRegular();
assertSidebarIcon(COLLECTION_NAME, "folder");
```

Reply:
If we have ONE test that asserts on the `official-collection-marker`, there is really no added value in doing it every time we change the collection type:

```javascript
it("successfully changes the colllection from the official to standard", ()=> {
  editCollection();
  changeCollectionFromOfficialToRegular();
  cy.findByTestId("official-collection-marker").should("not.exist");
  assertSidebarIcon(COLLECTION_NAME, "folder"); // the same goes for this assertion - we don't have to repeat it in every single test
});
```

then later on this function can be much more simplified and reduced to this (tests will also run faster)
```javascript
function changeCollectionFromOfficialToRegular() {
  modal().within(() => {
    setOfficial(false);
    cy.button("Update").click();
  });
}
```

The next logical step is to update the `setOfficial()` function and use it directly, which renders `changeCollectionFromOfficialToRegular()` and its counter-part totally unnecessary:
```javascript
function setOfficial(official = true) {
  const isOfficialNow = !official;
  
  cy.findByLabelText("Regular").should(
    isOfficialNow ? "not.be.checked" : "be.checked",
  );
  
  cy.findByLabelText("Official").should(
    isOfficialNow ? "be.checked" : "not.be.checked",
  );
  
  cy.findByText(official ? "Official" : "Regular").click();
  
  cy.button("Update").click();
}

editCollection();
setOfficla(true|false);
```